keps/sig-node/127-user-namespaces/README.md

```diff
@@ -610,6 +675,26 @@ enhancement:
   CRI or CNI may require updating that component before the kubelet.
 -->
 
+#### Kubelet and Kube-apiserver skew
+
+The apiserver and kubelet feature gate enablement work fine in any combination:
+
+1. If the apiserver has the feature gate enabled and the kubelet doesn't, then the pod will show
+   that field and the kubelet will ignore it. Then, the pod is created without user namespaces.
+2. If the apiserver has the feature gate disabled and the kubelet enabled, the pod won't show this
+   field and therefore the kubelet won't act on a field that isn't shown. The pod is created without
+   user namespaces.
+
+The kubelet can still create pods with user namespaces if static-pods are configured with
+pod.spec.hostUsers and has the feature gate enabled.
+
+If the kube-apiserver doesn't support the feature at all (< 1.25), a pod with userns will be
+rejected.
```

Comment:
This is not true - kube-apiserver is not rejecting the pod, but rather is silently dropping this field.

Also - this isn't only about the release <1.25, but rather also when the feature gate is disabled.

Reply:
No, it is rejecting the pod. This is what happens when a field is known, for example:

```
$ kubectl apply -f mypod2.yaml 
Error from server (BadRequest): error when creating "mypod2.yaml": Pod in version "v1" cannot be handled as a Pod: strict decoding error: unknown field "spec.nonExistingField"
```

I added the `pod.spec.nonExistingField: true` to the pod.

Am I missing something? Or are we talking about different things?


Regarding the feature gate disabled, that is mentioned a few lines above: https://github.com/kubernetes/enhancements/pull/4439/commits/3766329491a71f33501ed417c8d34640258fffeb#diff-a9ca9fbce1538447b92f03125ca2b8474e2d875071f1172d2afa0b1e8cadeabaR684-R686

Reply:
This is strange - the whole point of having the dropDisabledFields pattern is to handle and drop the field if the FG is disabled.
@liggitt - can you please shed some light on it?

Reply:
For context, this is the apiserver as started with the local-up-cluster.sh

Reply:
1. kubectl opts into strict mode, where unknown fields are flagged / rejected. `kubectl apply ... --validate=false` shows what normal API clients (like controllers) would experience, where unknown fields are just dropped.
2. emulating the behavior where unknown fields are dropped on create when a feature is disabled is the standard approach taken.
3. *rarely*, for a security-related feature, we will reject attempts to use the new security-related field if the feature is disabled rather than dropping the field. The only one I can remember off the top of my head was the windows `hostProcess` field

Reply:
do we even need this sentence? 1.25 has been out of support for a while, and we'd also not support a 1.30->1.25 skew between kubelet and KAS

Reply:
> We aren't doing anything special in the apiserver for this field (pod.spec.hostUsers) on the apiserver.

Once you add the field, you need to start doing things with it in the apiserver when the feature is disabled.

Reply:
Updated to specify the kube-apiserver will drop the field if it's unknown

Reply:
how does it look @liggitt @wojtek-t